src/Microsoft.Diagnostics.Tools.RuntimeClient/DiagnosticsIpc/IpcClient.cs

```diff
@@ -34,7 +34,11 @@ private static Stream GetTransport(int processId)
             {
                 string pipeName = $"dotnet-diagnostic-{processId}";
                 var namedPipe = new NamedPipeClientStream(
-                    ".", pipeName, PipeDirection.InOut, PipeOptions.None, TokenImpersonationLevel.Impersonation);
+                    ".", pipeName, PipeDirection.InOut, PipeOptions.None, TokenImpersonationLevel.Impersonation)
+                    {
+                        ReadTimeout = (int)ConnectTimeoutMilliseconds,
+                        WriteTimeout = (int)ConnectTimeoutMilliseconds
```

Comment:
This makes me nervous ... as a user of the tool if I have dotnet-trace running and a spurious timeout occurs have I just corrupted the recorded stream and lost the data? If I was in the middle of a long-running experiment that I wasn't closely monitoring would I come back the next day to find that most of the run was unrecorded and the experiment needs to be repeated?

Let me know if I am misreading this, but it appears to open up that class of issue. If so the cure might be worse than the disease if we aren't careful about how we design and test this.

Reply:
That's a good point.  Would it make more sense then too have some kind of heartbeat type logging mechanism to indicate when it is hung on a `read`?

Reply:
That's not something I considered - Thanks for pointing that out Noah! 

Quick thoughts though... Wouldn't we lose data from the runtime side anyway if the stream was hung at reading off the IPC and the runtime is accumulating events?

Reply:
There are "hangs" that we don't want to timeout, e.g., a non-verbose event type fires at a frequency longer than the timeout time of the read.  If this "hang" times out then `EventPipeEventSource` would end up with a prematurely closed handle on the transport and throw an exception right?  The true hang we want to timeout on is IPC responses.  Maybe we should put a timeout specifically on IPC read/writes and throw an exception if it times out.

Reply:
> Maybe we should put a timeout specifically on IPC read/writes and throw an exception if it times out

I think this is what we want to do. I'll modify the PR to do this. Noah, please let us know if this isn't the direction you want us to head :)

Reply:
I would have similar concerns about any situation where the client library has already connected, some work is in progress, and then the tool makes a decision to disconnect without any explicit action from the user. Typically the tools aren't in a position to understand whether a communication delay is meaningful or not. For example some potential causes of delay where the user probably wouldn't want the tool to disconnect:
a) Major machine resource starvation, app becomes temporarily unresponsive
b) Debugger or dump tool suspends the app for a while, then resumes it
c) VM pauses, migrates to a new host, then resumes. Timeout is incorrectly triggered when the real-time clock suddenly jumps forward to re-sync.
d) Request processing requires a lot of work so the response comes after considerable delay

Once any communication has occured it gets much more complicated to reason about the consequences of disconnect/retry vs. continuing to wait. I suspect the safe path is to design a UI interaction that provides the user status and lets the user decide when they want to disconnect. In some subset of cases it may also make sense to let the user specify a timeout or have a tool automatically set a timeout, but the RuntimeClient library probably isn't in position to make that judgement. Arguably we have a crude version of this right now, the status indicator is the user sees no progress and then the user decides to cancel by using ctrl-C. I'm sure we could do better, but the current state doesn't seem horrible to me, especially if such delays are extremely rare in practice?

Reply:
I agree that we should not be timing out once we have made the connection, but are you also saying that we shouldn't be timing out on the initial IPC message?

Reply:
> are you also saying that we shouldn't be timing out on the initial IPC message?

If you can guarantee that the message hasn't been received by the runtime I think a timeout would be OK, but I don't know how we'd be in position to assert that?

Reply:
For example if you sent a message to capture a GC dump lets assume the runtime starts the work but doesn't promptly respond to the request. The tool triggers a timeout and the user sees an error. If the user tries again the same thing will happen. Certainly it wasn't good that the runtime responded so slowly but the timeout is only exacerbating the problem by not allowing the user to ever get the result of their request.

Reply:
I think that's a fair point.  Let's say `dotnet-dump` all of a sudden takes 5 seconds to take the dump.  The IPC response for that command isn't sent till the dump is created and a 3 second timeout would throw an exception before the dump was done.  Maybe we should put a heartbeat message or similar in for long blocking IPC responses.  I don't feel like we would need that with an active EventPipe streaming session, but the initial IPC messages might benefit from some indication that this is what we are waiting on.

Reply:
Ok - I'll leave out doing this for now until I find out a better way to diagnose/deal with the runtime responding so slowly :-)

Reply:
The runtime isn't going to be in control of all the things that cause a slow response, at best we will only be able to mitigate some causes. I suggest moving the goal-posts : ) I wouldn't think about a timeout being the end goal, instead the goal is to give the user any information we have about whether progress is being made and let them make a decision. At the level of RuntimeClient this probably means providing a cancellation API or cancellation token.